### PR TITLE
2.x.x - sessions

### PR DESCRIPTION
**Pull request by adam-fowler**
- Merge three sessions examples into one. 

Things to note: 
- This is still using unmerged code from https://github.com/hummingbird-project/hummingbird-fluent/pull/11
- I couldn't use the `User` type as an `HBAuthenticatable` as Fluent model types cannot be set to `Sendable`.

### CODE REVIEW

File: sessions/Package.swift [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
     products: [
         .executable(name: "App", targets: ["App"]),
     ],
     dependencies: [
-        .package(url: "https://github.com/hummingbird-project/hummingbird.git", from: "1.8.0"),
-        .package(url: "https://github.com/hummingbird-project/hummingbird-auth.git", from: "1.2.0"),
-        .package(url: "https://github.com/hummingbird-project/hummingbird-fluent.git", from: "1.1.0"),
+        .package(url: "https://github.com/hummingbird-project/hummingbird.git", branch: "2.x.x"),
+        .package(url: "https://github.com/hummingbird-project/hummingbird-auth.git", branch: "2.x.x"),
+        .package(url: "https://github.com/hummingbird-project/hummingbird-fluent.git", branch: "2.x.x-working"),
```
Comment (Joannis) [inline sessions/Package.swift:15]: Reminder: This should be updated.

File: sessions/Sources/App/Application+build.swift [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    // routes
+    router.get("/") { _, _ in
+        return "Hello"
+    }
+
+    let userController = UserController(fluent: fluent, sessionStorage: sessionStorage)
+    userController.addRoutes(to: router.group("user"))
+
+    var application = HBApplication(responder: router.buildResponder(), server: .http1())
```
Comment (Joannis) [inline sessions/Sources/App/Application+build.swift:60]:
```suggestion
    var application = HBApplication(router: router, server: .http1())
```

File: sessions/Sources/App/Models/User.swift [hunk 3]
```diff
@@ -45,8 +45,21 @@ final class User: Model, HBAuthenticatable {
...
 }
 
+/// User type to pass around as authenticatable. This is required as Fluent
+/// model types are not Sendable
+struct LoggedInUser: HBAuthenticatable {
+    let id: UUID
+    let name: String
+
+    init(from user: User) throws {
+        guard let id = try? user.requireID() else { throw HBHTTPError(.unauthorized) }
```
Comment (Joannis) [inline sessions/Sources/App/Models/User.swift:55]: Any reason not to re-throw the error from `requireID()` here?

File: sessions/Sources/App/Models/User.swift [hunk 4]
```diff
@@ -58,16 +71,17 @@ struct CreateUserRequest: Decodable {
...
-    internal init(id: UUID?, name: String) {
+    internal init(id: UUID, name: String) {
         self.id = id
         self.name = name
     }
 
-    internal init(from user: User) {
-        self.id = user.id
+    internal init(from user: User) throws {
+        guard let id = try? user.requireID() else { throw HBHTTPError(.unauthorized) }
```
Comment (Joannis) [inline sessions/Sources/App/Models/User.swift:83]: I propose re-throwing the error from Fluent's `requireID()` here

**Review by Joannis — CHANGES_REQUESTED** [review]

File: sessions/Sources/App/Models/User.swift (hunk 3 above)
Comment (adam-fowler) [reply sessions/Sources/App/Models/User.swift:55]: If we throw the error from requireID the server will return a 500 as it doesn't know how to convert that into a usable error. Maybe it should return a 500 as user really should have an ID

File: sessions/Sources/App/Models/User.swift (hunk 4 above)
Comment (adam-fowler) [reply sessions/Sources/App/Models/User.swift:83]: see above

File: sessions/Sources/App/Models/User.swift (hunk 3 above)
Comment (adam-fowler) [reply sessions/Sources/App/Models/User.swift]: I went with your suggestion in the end

File: todos-dynamodb/Sources/App/Controllers/TodoController.swift [hunk 5]
```diff
@@ -71,7 +71,7 @@ struct TodoController: Sendable {
     }
 
     @Sendable func updateId(_ request: HBRequest, context: some HBRequestContext) async throws -> Todo {
-        guard var todo = try? await request.decode(as: EditTodo.self, context: context) else { throw HBHTTPError(.badRequest) }
+        var todo = try await request.decode(as: EditTodo.self, context: context)
         guard let id = context.parameters.get("id", as: UUID.self) else { throw HBHTTPError(.badRequest) }
```
Comment (Joannis) [inline todos-dynamodb/Sources/App/Controllers/TodoController.swift:75]: I like `require` more than `get` in these cases, what do you think?

File: sessions/Sources/App/Controllers/UserController.swift [hunk 6]
```diff
@@ -29,44 +29,46 @@ struct UserController {
...
+        return try UserResponse(from: user)
     }
 
     /// Login user and create session
-    func login(_ request: HBRequest) async throws -> HTTPResponseStatus {
+    @Sendable func login(_ request: HBRequest, context: SessionsContext) async throws -> HBResponse {
         // get authenticated user and return
-        guard let user = request.authGet(User.self),
-              let userId = user.id else { throw HBHTTPError(.unauthorized) }
+        guard let user = context.auth.get(LoggedInUser.self) else { throw HBHTTPError(.unauthorized) }
```
Comment (Joannis) [inline sessions/Sources/App/Controllers/UserController.swift:62]: Does `auth` have a `require`?

Comment (adam-fowler) [reply sessions/Sources/App/Controllers/UserController.swift:62]: Yeah if I remember right it does

File: todos-dynamodb/Sources/App/Controllers/TodoController.swift (hunk 5 above)
Comment (adam-fowler) [reply todos-dynamodb/Sources/App/Controllers/TodoController.swift]: yep fixed all the ones in this file

File: sessions/Sources/App/Controllers/UserController.swift (hunk 6 above)
Comment (adam-fowler) [reply sessions/Sources/App/Controllers/UserController.swift]: and fixed

**Review by Joannis — APPROVED** [review]

**Pull request merged**